--- ath6kl.ori/core.h	2012-11-15 22:39:08.000000000 +0800
+++ ath6kl/core.h	2012-11-16 16:27:10.000000000 +0800
@@ -444,8 +444,8 @@
 #define AGGR_TX_MAX_AGGR_SIZE   1600 /* Sync to max. PDU size of host. */
 #define AGGR_TX_MAX_PDU_SIZE    120
 #define AGGR_TX_MIN_PDU_SIZE    64   /* 802.3(14) + LLC(8) + IP/TCP(20) = 42 */
-#define AGGR_TX_MAX_NUM		6
-#define AGGR_TX_TIMEOUT         4	/* in ms */
+#define AGGR_TX_MAX_NUM		16
+#define AGGR_TX_TIMEOUT         4	/* in ms */
 
 #define AGGR_GET_TXTID(_p, _x)           (&(_p->tx_tid[(_x)]))
 
