--- NetworkManager-1.4.0/src/devices/wifi/nm-device-wifi.c.orig	2016-09-06 13:38:09.594338340 +0800
+++ NetworkManager-1.4.0/src/devices/wifi/nm-device-wifi.c	2016-09-06 17:11:37.208663792 +0800
@@ -1480,6 +1480,15 @@ schedule_scan (NMDeviceWifi *self, gbool
 	NMDeviceWifiPrivate *priv = NM_DEVICE_WIFI_GET_PRIVATE (self);
 	gint32 now = nm_utils_get_monotonic_timestamp_s ();
 
+	if (nm_device_get_state (NM_DEVICE (self)) == NM_DEVICE_STATE_ACTIVATED) {
+		if (priv->pending_scan_id) {
+			_LOGD (LOGD_WIFI_SCAN, "WiFi is already activated, cancel previous scheduled scan!");
+			nm_clear_g_source (&priv->pending_scan_id);
+		}
+		_LOGD (LOGD_WIFI_SCAN, "WiFi is already activated, do not schedule scan to reduce WiFi load!");
+		return;
+	}
+
 	/* Cancel the pending scan if it would happen later than (now + the scan_interval) */
 	if (priv->pending_scan_id) {
 		if (now + priv->scan_interval < priv->scheduled_scan_time)
