diff -up bind-9.9.1-P2/contrib/dlz/config.dlz.in.libdb4 bind-9.9.1-P2/contrib/dlz/config.dlz.in
--- bind-9.9.1-P2/contrib/dlz/config.dlz.in.libdb4	2012-07-30 16:58:57.566418514 +0200
+++ bind-9.9.1-P2/contrib/dlz/config.dlz.in	2012-07-30 17:30:10.930074108 +0200
@@ -257,7 +257,7 @@ case "$use_dlz_bdb" in
 			# Check other locations for includes.
 			# Order is important (sigh).
 
-			bdb_incdirs="/ /db48/ /db47/ /db46/ /db45/ /db44/ /db43/ /db42/ /db41/ /db4/ /db/"
+			bdb_incdirs="/ /db48/ /db47/ /db46/ /db45/ /db44/ /db43/ /db42/ /db41/ /db4/ /libdb4/ /db/"
 			for d in $bdb_incdirs
 			do
 				if test -f "$dd/include${d}db.h"
@@ -281,15 +281,9 @@ case "$use_dlz_bdb" in
 			bdb_libnames="db48 db-4.8 db47 db-4.7 db46 db-4.6 db45 db-4.5 db44 db-4.4 db43 db-4.3 db42 db-4.2 db41 db-4.1 db"
 			for d in $bdb_libnames
 			do
-				if test -f "$dd/${target_lib}/lib${d}.so"
+				if test -f "$dd/${target_lib}/libdb4/lib${d}.so"
 				then
-					if test "$dd" != "/usr"
-					then
-						dlz_bdb_libs="-L${dd}/${target_lib} "
-					else
-						dlz_bdb_libs=""
-					fi
-					dlz_bdb_libs="${dlz_bdb_libs}-l${d}"
+					dlz_bdb_libs="-L${dd}/${target_lib}/libdb4 -l${d}"
 					break
 				fi
 			done
