--- ath6kl.orig/debug.c	2012-07-12 16:19:02.000401000 +0800
+++ ath6kl/debug.c	2012-07-12 16:50:14.195445229 +0800
@@ -42,16 +42,23 @@
 
 int ath6kl_printk(const char *level, const char *fmt, ...)
 {
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,36))
 	struct va_format vaf;
+#endif
 	va_list args;
 	int rtn;
 
 	va_start(args, fmt);
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,36))
 	vaf.fmt = fmt;
 	vaf.va = &args;
 
 	rtn = printk(KERN_INFO "%sath6kl: %pV", level, &vaf);
+#else
+	printk("%sath6kl: ", level);
+	rtn = vprintk(fmt, args);
+#endif
 
 	va_end(args);
 
