diff -Naur ath6kl.ori/core.h ath6kl/core.h
--- ath6kl.ori/core.h	2012-12-06 17:52:22.000000000 +0800
+++ ath6kl/core.h	2012-12-06 19:23:55.000000000 +0800
@@ -178,7 +178,7 @@
 #define ATH6KL_ROC_MAX_PERIOD		(5)	/* in sec. */
 
 /* scan time out */
-#define ATH6KL_SCAN_TIMEOUT_LONG (8 * HZ)  /* in sec. */
-#define ATH6KL_SCAN_TIMEOUT_SHORT (5 * HZ) /* in sec. */
+#define ATH6KL_SCAN_TIMEOUT_LONG (20 * HZ)  /* in sec. */
+#define ATH6KL_SCAN_TIMEOUT_SHORT (15 * HZ) /* in sec. */
 #define ATH6KL_SCAN_TIMEOUT_WITHOUT_ROAM (20 * HZ)  /* in sec. */

 /* 4 way-handshake protect */
